### PR TITLE
Integrate pre-commit tool

### PR DESCRIPTION
**Pull request by webknjaz**
pre-commit integration distilled from #23

### CODE REVIEW

**Comment by pradyunsg** [comment]
I'll defer to others on whether we want this. :)

**Comment by webknjaz** [comment]
@pradyunsg mind tagging them?

File: 2.6/get-pip.py [hunk 1]
```diff
@@ -135,7 +135,7 @@ def parse_args(self, args):
     for arg in args:
         try:
             req = InstallRequirement.from_line(arg)
-        except:
+        except Exception:
```
Comment (pfmoore) [inline 2.6/get-pip.py]: Unrelated change?

File: 3.2/get-pip.py [hunk 2]
```diff
@@ -135,7 +135,7 @@ def parse_args(self, args):
     for arg in args:
         try:
             req = InstallRequirement.from_line(arg)
-        except:
+        except Exception:
```
Comment (pfmoore) [inline 3.2/get-pip.py]: Unrelated change?

File: 3.3/get-pip.py [hunk 3]
```diff
@@ -135,7 +135,7 @@ def parse_args(self, args):
     for arg in args:
         try:
             req = InstallRequirement.from_line(arg)
-        except:
+        except Exception:
```
Comment (pfmoore) [inline 3.3/get-pip.py]: Unrelated change?

File: templates/default.py [hunk 4]
```diff
@@ -135,7 +135,7 @@ def parse_args(self, args):
     for arg in args:
         try:
             req = InstallRequirement.from_line(arg)
-        except:
+        except Exception:
```
Comment (pfmoore) [inline templates/default.py]: Unrelated change?

File: templates/pre-10.py [hunk 5]
```diff
@@ -135,7 +135,7 @@ def parse_args(self, args):
     for arg in args:
         try:
             req = InstallRequirement.from_line(arg)
-        except:
+        except Exception:
```
Comment (pfmoore) [inline templates/pre-10.py]: Unrelated change?

**Comment by pfmoore** [comment]
The documentation on `pre-commit` was a bit hard to find (the PyPI description was "UNKNOWN" and the github readme was pretty minimal, although it did point to a project website, which looks extensive). One thing I did spot that bothers me a lot: "does not work on platforms without symlink support (windows)". I'm a strong -1 on including anything in the workflow that doesn't work on Windows.

Also, I'd want to see some documentation added here that explained how to use this. I don't want to direct project maintainers to go and read the whole of https://pre-commit.com/ to understand what's going on.

Also, I'm not a huge fan of mandatory pre-commit hooks. If I'm just making a trivial change (such as docs) I will often do so in an environment that doesn't have any dev requirements installed, on the expectation that CI will act as the fallback check. So I'd rather we have proper CI here before looking at mandating a certain setup on developer machines.

Overall, I'd need to be convinced of the benefits of this change before approving it.

File: 2.6/get-pip.py (hunk 1 above)
Comment (webknjaz) [reply 2.6/get-pip.py]: It's related, cherry-picked from #23. Some of linters (flake8) was yelling at it. I think, adding linters should go along with cleaned code.

**Comment by webknjaz** [comment]
@pfmoore 
> I'm a strong -1 on including anything in the workflow that doesn't work on Windows.

I guess, it depends on linters/tools you add to config, but it works in general (@asottile, am I right?):

> Support: python hooks work without any system-level depedendencies. It has been tested on linux, macOS, windows, and cygwin.

---
> Also, I'm not a huge fan of mandatory

I didn't say it's mandatory. It's more like "use this command **if you want** to automatically run the same locally." If it's not clear from wording, the explanation should be fixed.

---
> that CI will act as the fallback check.

We were trying to do this in #23, but @pradyunsg asked to have a separate PR for this. (https://github.com/pypa/get-pip/pull/23/commits/2559f917379648f1f525fcc34eb4644c5d16aec3)

---
> benefits of this change

I personally like that it (1) allows to consolidate several linters in one config and (2) be able to run this in CI. Also (3) I like that there's possibility to automagically install it into your local repo (pre-commit or pre-push hook), which runs in special mode taking into account only changed files (as opposed to `--run-all` in CI).

After accepting this and #23 I can contribute moving it all into tox for example.

---
> Also, I'd want to see some documentation

I'm going to do this once the general idea of having this tooling is accepted.

File: 2.6/get-pip.py (hunk 1 above)
Comment (pfmoore) [reply 2.6/get-pip.py]: If they are fixes for linter issues, can they be put in a separate PR, please?

Comment (webknjaz) [reply 2.6/get-pip.py]: @pfmoore Done!

Comment (webknjaz) [reply 2.6/get-pip.py]: #30

File: README.rst [hunk 6]
```diff
@@ -36,6 +36,13 @@ You may want to perform something like:
     $ source venv/bin/activate
     $ pip install -r requirements.txt
 
+It is also recommended to make linters run to keep up with coding standards
```
Comment (webknjaz) [inline README.rst]: @pfmoore any suggestion on wording improvements?

**Comment by pfmoore** [comment]
> I didn't say it's mandatory. It's more like "use this command if you want to automatically run the same locally."

Well, the way we do the test suite in pip (tests in py.test, run using tox, integrated into CI using simple Appveyor and Travis configs that just run tox) have the same benefit, and additionally, the arrangement is familiar to people who work on pip.

I'm not against trying out a new tool, but I personally have no experience with it and I've never seen any other project that uses it - so for me, at least, it has to prove its advantages over more standard approaches.

One other point, which I made in #18, is that we should test building and running `get-pip` from the *pip* tests. Changes in get-pip itself are very rare, and typically don't cause problems. The problems arise when changes in pip cause issues with get-pip, so I'd prefer to see the tests happening over there. That's not to say that we can't also have tests in get-pip itself, just that I don't think they really address the need in #18.

> I'm going to do this once the general idea of having this tooling is accepted.

Chicken and egg situation here. Without docs on how this would affect our workflow, it's hard to know if this approach is acceptable. But it's your call - I appreciate that writing docs is a pain, particularly if there's a chance the work will be wasted. Maybe you could just give examples of some other Python projects that use this tool, that we could look at for reference?

> We were trying to do this in #23, but @pradyunsg asked to have a separate PR for this. (2559f91)

In which case I'm confused, as the checks in here don't seem to be the same as the ones in #23. The fact that I have difficulty confirming that reinforces my feeling that maintaining these tests will be hard to do.  

Also, the checks you're adding seem like a very odd list. Why is there a `check-yaml` test - we don't have any YAML files? Plus, by deferring the actual details of the checks to a 3rd party implementation, we lose control over what we test.

These tests all seem to be "lint" style tests, which aren't really that important (IMO) for get-pip, which is a very small, and unusual style of codebase (most of the code is an opaque binary blob). I'm not against code style checkers, but I'm frankly not that interested with adding them here.

File: README.rst (hunk 6 above)
Comment (pfmoore) [reply README.rst]: No, not really. As I said, I'm not convinced we need linter checks on this project, so I'm hardly the one to say how you should word a recommendation that we use them :smile:

**Comment by pfmoore** [comment]
>> My apologies for the comment on Windows support
>
> No problem :) I've invested an absolute ton of work to support windows, I guess I should call it out a bit more!

You should actually just assume your readers aren't paranoid :smile: I suspect it's only refugees from the old OS wars like me that still assume the worst. Nearly every project I encounter these days either is, or is trying to be, cross-platform. You'd think I would have learned by now...

**Comment by webknjaz** [comment]
@pfmoore yes, there's a few hooks which are not applicable and are artifacts of copy-pasting stuff :) (yaml and probably a few of others)

Here's how we invoke it in tox: https://github.com/cherrypy/cheroot/blob/master/tox.ini#L30-L37
There's a separate env for running linters, which currently generate errors (for gradual fixing).

Thanks @asottile for those explanations btw ;)

With regard to whether linters are needed, it's simple: who wants


Oh, and regarding templates: during work on #23 we had flake8 lint those and it didn't seem to cause problems.

**Comment by pfmoore** [comment]
This is getting a bit philosophical, so I don't intend to debate this much further, but

> With regard to whether linters are needed, it's simple: who wants their code to be diverse? Also linters help to prevent accidental errors, so why not?

The "why not" is simple - because it encourages extra "keep the linter happy" commits. On most projects, that's essentially irrelevant, and the benefit of a consistent style (and an impartial arbiter to avoid "I prefer it this way" arguments) over a large codebase is worth it. On this project, where we essentially have only one Python file, and every commit to master is a release, and where most of our commits are direct to master, those extra commits are more of a concern. Not to the extent of being a showstopper, but enough that I'm personally inclined to be super-cautious. The other committers (particularly @dstufft who knows much better than me how this repo works) may feel differently.

**Comment by webknjaz** [comment]
> encourages extra "keep the linter happy" commits

Yes, but it might as well be non-mandatory, Travis CI allows making jobs allowed to fail, which will not fail the whole build, but will present the current status to everyone interested. From this point of view, it comes for no cost. I personally prefer if project I contribute to sets some style expectations and linters are the right tools for this, since it shouldn't be on humans.

I think, the most debatable thing is which linters to use, I've provided some example with this PR, but it doesn't mean that it's the best collection of checks possible.

Anyway, there's other things to contribute, like tox and CI/CD improvements, so I think this PR should not be blocker for those.

**Comment by pradyunsg** [comment]
I agree with @pfmoore. This project doesn't need linter checks -- we basically only regenerate get-pip.py scripts and publish them here, and at this point, there's no active development / changes being made here that would benefit from having consistent code style etc. We can revisit this if we add a bunch more code here but I think I'll go ahead and close this for now. :)

Thanks for the PR @webknjaz!
